README
说明:
这是Meteor Liu的emacs配置文件，兼容Linux/Windows/MacOSX下的emacs-22以上版本，我用emacs主要进行C++开发和用org管理TODO，有时候也写点python。
init-basic.el是基本的emacs配置，只需要emacs自身，不依赖第三方插件。
init-misc.el用于配置lisp目录下的第三方插件。
init-site.el用于配置不在lisp目录下的其它第三方插件，比如cedet，ecb，jdee，pymacs等，我一般把它存在site-lisp(Windows)或~/.emacs.d(Linux and MacOSX)目录下，主要是因为这些插件比较大，我不愿意把它存在git里。
init-toolbar.el是我的工具栏配置。
lisp目录是第三方elisp插件。
site-lisp目录是本地elisp目录(未包含在版本库里，用于本地多操作系统共享)。
bin目录用来存可执行文件，比如win32下默认没有的grep等。
etc/snippets目录是yasnippet的snippets定义。
etc/templates目录是autoinsert模板。
etc/images目录是toolbar用的图标文件。
etc/sample/emacs是.emacs文件示例。
etc/sample/proj.el是配置ede-cpp-root-project(参考cedet的ede)项目的示例文件。
etc/sample/diary是diary文件的例子。
etc/sample/site-start.el是portable emacs的例子，把这个文件拷贝到site-lisp目录就行。
etc/sample/dir-locals.el是.dir-locals.el的例子。

安装:
git clone http://github.com/meteor1113/dotemacs.git
把etc/sample/emacs拷贝到$HOME目录下改名为.emacs，修改load-path为dotemacs的实际路径，根据需要可注释掉部分内容。
如果需要cedet，可以从 http://cedet.sourceforge.net 下载cedet并安装（Windows我一般直接解压在site-lisp目录下，Linux和MacOSX一般解压到~/.emacs.d目录下）。
如果需要ecb，可以从 http://ecb.sourceforge.net 下载ecb并安装，安装位置可参考cedet。
如果需要用jdee开发java，可以从 http://jdee.sourceforge.net 下载jdee并安装，安装位置可参考cedet。
如果需要用ropemacs开发python，可以分别从 http://rope.sourceforge.net ， http://pymacs.progiciels-bpi.ca/pymacs.html ， http://rope.sourceforge.net/ropemacs.html 下载rope，pymacs，ropemacs并安装，把pymacs编译时生成的pymacs.el文件放在site-lisp或~/.emacs.d目录下。（注意ropemacs需要下载snapshot版，release的版本不能补全）

Email: meteor1113@gmail.com


ChangeLog
2020-09-09  Liu Xin  <meteor1113@qq.com>

	* init/init-editor.el (global-tab-line-mode): use tab-line-mode to
	instead of tabbar-mode.
	(t): use (save-place-mode) instead of (setq save-place t)

2018-06-08  Liu Xin  <meteor1113@qq.com>

	* init/init-editor.el (fboundp): Use display-line-numbers-mode in emacs26.

2018-04-22  Liu Xin  <meteor1113@qq.com>

	* init/init-editor.el (whitespace-style): Adjust whitespace.

2017-11-01  Liu Xin  <meteor1113@qq.com>

	* init/modules/init-ecb.el : Change ecb-compile-window-height to 0.125

2017-10-13  Liu Xin  <meteor1113@qq.com>

	* init/modules/init-font.el: Upgrade chinese-fonts-setup to cnfonts.

2017-08-07  Liu Xin  <meteor1113@qq.com>

	* init/modules/init-toolbar.el: Fix toolbar for osx's crash.

2017-04-15  Liu Xin  <meteor1113@qq.com>

	* init/init-keybinding.el (eq): Bind command-<left> and
	command-<right> to home and end in os x.
	(eq): Key binding for os x.

2017-04-07  Liu Xin  <meteor1113@qq.com>

	* init/init-editor.el (after-init-hook): Disable zone for performance.

2017-03-21  Liu Xin  <meteor1113@qq.com>

	* init/modules/init-toolbar.el (eq): Disable toolbar for osx.

2017-01-23  Liu Xin  <meteor1113@qq.com>

	* init/init-package.el: Add package editorconfig.

2016-09-19  Liu Xin  <meteor1113@qq.com>

	* init/modules/init-font.el: Add init-font.el
	Improve init-font.el

2016-08-31  Liu Xin  <meteor1113@qq.com>

	* init/modules/init-tabbar.el: Don't load tabbar-ruler in daemon.

2016-07-29  Liu Xin  <meteor1113@qq.com>

	* init-emacs.el: Rename init.el to init-emacs.el

	* init/init-package.el: Use after-init-hook instead of run-with-idle-timer.

2016-04-27  Liu Xin  <meteor1113@qq.com>

	* init/init-editor.el (frame-title-format): Modify frame-title-format.

2016-02-06  Liu Xin  <meteor1113@qq.com>

	* init/modules/init-cedet.el ("pulse"): Speed up pulse.

2016-01-29  Liu Xin  <meteor1113@qq.com>

	* init/modules/init-toolbar.el: Add minor mode for helm-mode.

	* init/modules/init-calendar.el: Update appt setting for diary.

2016-01-28  Liu Xin  <meteor1113@qq.com>

	* init.el: Speed up.

2016-01-27  Liu Xin  <meteor1113@qq.com>

	* init/init-editor.el: Disable global-cwarn-mode.
	Don't load generic-x at start up.

2016-01-26  Liu Xin  <meteor1113@qq.com>

	* init/modules/init-toolbar.el: Update major mode menu for haskell.

	* init/progmodes/init-haskell.el: Add package ghc and company-ghc.

	* init/modules/init-fci.el: Config fill-column-indicator.

	* init/progmodes/init-prog.el: Only enable fci-mode in window-system.

	* init/modules/init-flycheck.el: Disable global-flycheck-mode default.

	* init/init-package.el (fci-dash-pattern): Config fci-mode and whitespace.

2016-01-25  Liu Xin  <meteor1113@qq.com>

	* init/init-package.el: Disable symon-mode default.

	* init/modules/init-toolbar.el: Update toolbar menu.

2016-01-24  Liu Xin  <meteor1113@qq.com>

	* init/init-package.el: Add package minimap.
	(after-init-hook): Add package symon.
	(after-init-hook): Add package aggressive-indent.

	* init/progmodes/init-gdb.el: Split init-cc to init-cc and init-gdb.

	* init/init-package.el (after-init-hook): Add pacage smex.

2016-01-23  Liu Xin  <meteor1113@qq.com>

	* init/init-package.el: Add package anzu.

2016-01-20  Liu Xin  <meteor1113@qq.com>

	* init/init-editor.el: Keep desktop-load-locked-desktop to 'ask.

	* init.el: Don't server-start automatic.

2016-01-19  Liu Xin  <meteor1113@qq.com>

	* init/modules/init-cedet.el (featurep): Fix called-interactively-p.

	* init.el: Add 'noerror for require.

	* init/modules/init-highlight.el (after-init-hook): Setting for volatile-highlights.

	* init/modules/init-tabbar.el: Update tabbar setting for tabbar-ruler-20160115.923

	* init/textmodes/init-org.el ("org"): Fix for yasnippet.

2016-01-09  Liu Xin  <meteor1113@qq.com>

	* init/progmodes/init-haskell.el: Add init-haskell.

	* init/modules/init-toolbar.el: Remove flycheck-clear.xpm

2015-12-31  Liu Xin  <meteor1113@qq.com>

	* init/modules/init-toolbar.el: Set toolbar for flycheck.

	* init/modules/init-flymake.el: Remove flymake.

	* init/modules/init-flycheck.el: Add flycheck.

2015-12-30  Liu Xin  <meteor1113@qq.com>

	* init/modules/init-toolbar.el: Remove recent-jump toolbar.

	* init/modules/init-helm.el: Add helm.

	* init/modules/init-toolbar.el (toggle-toolbar-menu): Refactor init-toolbar.

2015-12-28  Liu Xin  <meteor1113@qq.com>

	* init.el: Refactor.

	* init/init-editor.el: Remove htmlize.
	Remove cursor-chg.
	Remove recent-jump.

	* init.el: Split dotemacs.
	Refactor.

2015-12-26  Liu Xin  <meteor1113@qq.com>

	* init-lisp.el: Remove sourcepair.

	* init-basic.el (c-mode-common-hook): Setup ff-find-other-file.

2015-12-25  Liu Xin  <meteor1113@qq.com>

	* init-lisp.el: Remove eim.

	* init-elpa.el: Remove anything.
	("chinese-wbim"): Add chinese-wbim.

	* init-lisp.el: Remove codepilot.

	* init-basic.el (highlight-text): Add highlight-text. Remove smart-hl.

2015-12-24  Liu Xin  <meteor1113@qq.com>

	* init-lisp.el: Remove xgtags.
	Remove winsav.
	Remove word-count.
	Remove vimpulse.
	Remove ifdef.
	Remove code-imports.
	Remove cn-weather.
	Remove doxymacs and doc-mode.

	* init-elpa.el ("rust-mode"): Config rustfmt, racer, ac-racer.
	(rust-mode-hook): Add cargo.
	Add go-mode, go-autocomplete, go-doc, go-snippets.
	: Add ac-clang, Remove auto-complete-clang.
	: Add ggtags.

2015-12-23  Liu Xin  <meteor1113@qq.com>

	* init-elpa.el: Add rust-mode, rustfmt, racer, ac-racer.

2015-09-08  Liu Xin  <meteor1113@qq.com>

	* init-elpa.el (ecb-symboldef-find-tag-by-etags): Disable etags's
	'Visit tags tagle' dialog when ecb's symboldef is actived.

2015-06-10  Liu Xin  <meteor1113@qq.com>

	* init-basic.el (format-xml-directory): Add function format-xml-directory.

2015-05-28  Liu Xin  <meteor1113@qq.com>


init-misc.el
;;; -*- mode: emacs-lisp; mode: goto-address; coding: utf-8; -*-
;; Copyright (C) 2008-2015 Liu Xin
;;
;; This code has been released into the Public Domain.
;; You may do whatever you like with it.
;;
;; @file
;; @author Liu Xin <meteor1113@gmail.com>
;; @date 2008-08-08
;; @URL http://git.oschina.net/meteor1113/dotemacs


;; path
(let* ((dir (file-name-directory (or load-file-name (buffer-file-name))))
       (default-directory (expand-file-name "lisp" dir)))
  (when (file-exists-p default-directory)
    (add-to-list 'load-path default-directory)
    (if (fboundp 'normal-top-level-add-subdirs-to-load-path)
        (normal-top-level-add-subdirs-to-load-path))))

;; (unless (fboundp 'define-global-minor-mode) ; for emacs-21
;;   (defmacro define-global-minor-mode (global-mode mode turn-on &rest keys)))
(unless (fboundp 'define-globalized-minor-mode) ; for emacs-22
  (defalias 'define-globalized-minor-mode 'define-global-minor-mode))
(unless (fboundp 'with-no-warnings)     ; for emacs-21
  (defun with-no-warnings (body)
    "Before emacs-21, have not with-no-warnings function."))
(unless (fboundp 'define-fringe-bitmap) ; for emacs-21
  (defun define-fringe-bitmap (var value)
    "Before emacs-21, have not define-fringe-bitmap function."))

;; unicad
(require 'unicad nil 'noerror)

;; cal-china-x
(eval-after-load "calendar"
  '(when (require 'cal-china-x nil 'noerror)
     (setq cal-china-x-priority1-holidays
           (append holiday-local-holidays
                   cal-china-x-chinese-holidays
                   cal-china-x-japanese-holidays))
     (setq calendar-holidays
           (append calendar-holidays
                   cal-china-x-chinese-holidays
                   cal-china-x-japanese-holidays))))

;; cn-weather
;; (setq cn-weather-city "大连")
(autoload 'display-cn-weather-mode "cn-weather"
  "Display weather information in the mode line." t)
(autoload 'cn-weather "cn-weather"
  "Print Now today's and realtime weather in the echo area." t)
(autoload 'cn-weather-forecast "cn-weather"
  "Print future two days' weather info in minibuffer." t)

;; vimpulse
(setq viper-mode nil)
(eval-after-load "viper"
  '(require 'vimpulse nil 'noerror))

;; ascii
(autoload 'ascii-on        "ascii" "Turn on ASCII code display."   t)
(autoload 'ascii-off       "ascii" "Turn off ASCII code display."  t)
(autoload 'ascii-display   "ascii" "Toggle ASCII code display."    t)
(autoload 'ascii-customize "ascii" "Customize ASCII code display." t)

;; recent-jump
(when (require 'recent-jump nil 'noerror)
  (global-set-key (kbd "<M-S-left>") 'recent-jump-jump-backward)
  (global-set-key (kbd "<M-S-right>") 'recent-jump-jump-forward))

;; ace-jump-mode
(autoload 'ace-jump-mode "ace-jump-mode" nil t)
(define-key global-map (kbd "C-c SPC") 'ace-jump-mode)
(eval-after-load "ace-jump-mode"
  '(set-face-background 'ace-jump-face-foreground "yellow"))
(eval-after-load "viper-keym"
  '(define-key viper-vi-global-user-map (kbd "SPC") 'ace-jump-mode))

;; smart-hl
(when (> emacs-major-version 21)
  (require 'smart-hl nil 'noerror))

;; hl-defined
(autoload 'hdefd-highlight-mode "hl-defined" nil t)
(add-hook 'emacs-lisp-mode-hook
          '(lambda ()
             (when (require 'hl-defined nil 'noerror)
               (hdefd-highlight-mode 1)))
          'APPEND)

;; ifdef
(add-hook 'c-mode-common-hook
          '(lambda ()
             (when (require 'ifdef nil 'noerror)
               (define-key c-mode-base-map [?\C-c ?\C-i] 'mark-ifdef)
               (mark-ifdef))))

;; doc-mode/doxymacs-mode
(unless (locate-library "url")
  (provide 'url))                       ; emacs-21 doesn't have url
(add-hook 'c-mode-common-hook
          '(lambda ()
             (if (and (featurep 'semantic)
                      (require 'doc-mode nil 'noerror))
                 (doc-mode t)
               (when (require 'doxymacs nil 'noerror)
                 (doxymacs-mode t)
                 (doxymacs-font-lock)))))

;; xgtags
(when (executable-find "global")
  (add-hook 'c-mode-common-hook
            (lambda ()
              (when (require 'xgtags nil 'noerror)
                (xgtags-mode 1)))))

;; ntcmd
;; (autoload 'ntcmd-mode "ntcmd" "Major mode for editing CMD scripts." t)
;; (add-to-list 'auto-mode-alist '("\\.[bB][aA][tT]\\'" . ntcmd-mode))
;; (add-to-list 'auto-mode-alist '("\\.[cC][mM][dD]\\'" . ntcmd-mode))

;; sourcepair
(setq sourcepair-source-extensions
      '(".cpp" ".cxx" ".c++" ".CC" ".cc" ".C" ".c" ".mm" ".m"))
(setq sourcepair-header-extensions
      '(".hpp" ".hxx" ".h++" ".HH" ".hh" ".H" ".h"))
(setq sourcepair-header-path '("." "include" ".." "../include" "../inc"
                               "../../include" "../../inc" "../*"))
(setq sourcepair-source-path '("." "src" ".." "../src" "../*"))
(setq sourcepair-recurse-ignore '("CVS" ".svn" ".hg" ".git" ".bzr"
                                  "Obj" "Debug" "Release" "bin" "lib"))
(add-hook 'c-mode-common-hook
          '(lambda ()
             (when (require 'sourcepair nil 'noerror)
               (define-key c-mode-base-map (kbd "ESC <f12>") 'sourcepair-load)
               (define-key c-mode-base-map [M-f12] 'sourcepair-load))))

;; code-imports
(autoload 'code-imports-grab-import "code-imports" nil t)
(autoload 'code-imports-add-grabbed-imports "code-imports" nil t)
(autoload 'code-imports-organize-imports "code-imports" nil t)
(setq code-imports-project-directory ".")
(global-set-key (kbd "C-S-o") 'code-imports-organize-imports)
(eval-after-load "code-imports"
  '(defun code-imports--import-in-group-p (import-line
                                           group
                                           &optional self-file)
     "Returns t if IMPORT-LINE is in GROUP.
GROUP is one of the elements of the ordering such as
`code-imports-c++-ordering'. SELF-FILE is the .h file
corresponding to the file being modified (or nil if we're not in
a c mode)."
     (cond ((eq group 'self)
            ;; right now self can only refer to c/c++ mode.
            (string-match (regexp-quote (file-name-sans-extension self-file))
                          import-line))
           ;; Ignore the t group for these kinds of matches, otherwise
           ;; it won't match just things not matched by other groups.
           ;; t-matching will happen elsewhere.
           ((eq group t)
            nil)
           (t (string-match group import-line)))))

;; word-count
(autoload 'word-count-mode "word-count"
  "Minor mode to count words." t nil)
(global-set-key "\M-+" 'word-count-mode)

;; eim
;; (when (<= emacs-major-version 21)
;;   (provide 'help-mode)
;;   (defalias 'locate-file 'locate-library)
;;   (defvar emacs-basic-display nil))
(autoload 'eim-use-package "eim" "The eim input method" t)
(register-input-method
 "eim-wb" "euc-cn" 'eim-use-package "eim-wb" "eim-wb" "wb.txt")
(register-input-method
 "eim-py" "euc-cn" 'eim-use-package "eim-py" "eim-py" "py.txt")
(setq default-input-method "eim-wb")
;; (setq eim-use-tooltip nil)
(eval-after-load "eim"
  `(when (require 'eim-extra nil 'noerror)
     (global-set-key ";" 'eim-insert-ascii)))

;; (when (require 'winsav nil t)
;;   (winsav-save-mode 1))
(autoload 'winsav-save-configuration "winsav" nil t)
(autoload 'winsav-restore-configuration "winsav" nil t)

(provide 'init-misc)
